### ISSUE TITLE
Inconsistent behavior when reducing empty (Nil) expressions with comparison and boolean operators

### ISSUE THREAD

**Issue by wangjia184** [rchain/rchain#3344]
The following rholang code is correct

```
new somename, stdout(`rho:io:stdout`) in {
    somename!(Nil) |
    for(@x <- somename) {
        if( x != Nil and x.get("something") > 0 ) {
            stdout!("true")
        } else {
            stdout!("false")
        }
    }
}
```

### What did you expect to see?

It should output either `true` or `false`. In this specific example, it should print `false`

### What did you see instead?

Neither `true` nor `false` is printed, also it does not raise error 

### Steps to reproduce the bug

First start RNode in docker as below

```
docker run -d --name rnode --rm \
    -p 40401:40401 \
    -p 40402:40402 \
    -p 40403:40403 \
    -p 40404:40404 \
    -p 40405:40405 \
    -v $HOME/var/rnode1:/var/lib/rnode \
    rchain/rnode:latest run --standalone
```

then store the rholang code into `bug1.rho`, and execute:

```
#!/bin/bash
docker run --rm --name rnode-client \
    --network host -it \
    -v $HOME/rholang:/rholang:ro \
    rchain/rnode:latest \
    --grpc-port 40402 eval /rholang/bug1.rho
```

### What version of RNode are you using?

RChain Node 0.10.1 `b159823c503d945917ead913b12e07c3265bff78`

### What is your environment?
Linux docker 

After modifying the code as below, it works well.  
It seems the keyword `and` does not work properly.

```
new somename, stdout(`rho:io:stdout`) in {
    somename!(Nil) |
    for(@x <- somename) {
        if( x != Nil ) {
            if( x.get("something") > 0 ) {
                stdout!("true")
            } else {
                stdout!("false")
            }
            
        } else {
            stdout!("false")
        }
    }
}
```

**Comment by 9rb** [comment]
new stdout,somename in {
  somename!(1) |
  for(@x <- somename) {
    if( ( x != 1) and  (x.get("something") > 0) ) {
      stdout!("true")
    } else {
      stdout!("false")
    }
  }
}

gives

Error: Method `get` is not defined on Int.

But in case of Nil, we do not get the above error message but "Error: Multiple expressions given." I used https://rgov1.netlify.app/ to test this piece of code.

**Comment by tgrospic** [comment]
The problem is how reducer evaluates comparison and boolean operators. There are two places where error with the message `"Multiple expressions given."` can be thrown. @9rb what you reported is from here.
https://github.com/rchain/rchain/blob/09d9acc916e008dc93a7289202262fd5a4d9d41d/rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala#L1448-L1456

Here is another place. In both places, the case with empty list of expressions it's not handled so empty and more than one element will display mentioned error.
https://github.com/rchain/rchain/blob/09d9acc916e008dc93a7289202262fd5a4d9d41d/rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala#L1503-L1513

The first case, when error is thrown inside `evalSingleExpr`, can be reproduced with `true and Nil` or `Nil > 1` or `Nil.get("")`, but not equality e.g. `Nil == Nil` or `Nil != Nil` which are implemented differently.
Second case is with boolean operators (`evalToBool`) when operand is empty expression e.g. `Nil and true` or `Nil or Nil`.

To resolve this bug these questions needs to be answered.

1. Should we throw exception when evaluating empty expressions (`Nil`)?
1. Is it possible to apply a fix without causing hard-fork on the blockchain? At least error message can be fixed.

If error is not thrown on empty expressions both comparison and boolean operators will handle incompatible operators and throw corresponding errors.
`if("x" and 0) {}` _Error: expression didn't evaluate to boolean._
`if("x" > 0) {}` _Unexpected compare: Expr(GString(x)) vs. Expr(GInt(0))_

---

@wangjia184 Thanks for reporting this bug!! Is this description cover your observations or you think there is more to it?

**Comment by 9rb** [comment]
@tgrospic , 
1.  I think the originally reported bug is this : "A and B and .. n " should always evaluate to false if ANY ONE of the component expressions evaluates to false. (This of course applies ONLY to the 'false' case). In the example wangjja184 provided, the first component expression evaluates to false, but the "and" gets lost in the indeterminacy of the second component. It should not matter.
2. The error messages may need to be rationalized - e.g. perhaps both (a) the fact that .get became indeterminate and (b) whatever trouble the boolean evaluation had , should be reported ?
3. May be we should assess if .get can be defined (as opposed to indeterminate) on more cases ? But this can open a can of worms and cause huge testing burden.  Perhaps can be added to 'nice to have' list for now.

**Comment by tgrospic** [comment]
@9rb What I described is the root cause of all these cases you have mentioned.

Calling `x.get` on a process `x` which doesn't support it will throw exception. So the problem is again if `x` is Nil which trigger this bug before it evaluates `get` method.

**Comment by wangjia184** [comment]
> Calling x.get on a process x which doesn't support it will throw exception. So the problem is again if x is Nil which trigger this bug before it evaluates get method.

Yep, that is the reason I assumed.    
`and` branches should be evaluated from left to right. If the left branch returns false, the right branch do not need be evaluated.
`or` branches are similar, if left one returns true, the right branch do not need be evaluated.
Agree?

**Comment by 9rb** [comment]
Interestingly, getOrElse gives a more useful error message.  The following

new stdout,somename in {
  somename!(Nil) |
  for(@x <- somename) {
    if( x != Nil and x.getOrElse(0, ("something") > 0)) {
      stdout!("true")
    } else {
      stdout!("false")
    }
  }
}

gives

Result

[ ]

Problem

Unexpected compare: Expr(GString(something)) vs. Expr(GInt(0))

**Comment by tgrospic** [comment]
@9rb that's good observation, although it's not related to `.getOrElse` method defined on Rholang `Map`.

Comparisons operators (except equality `==`, `!=`) have check for type equality of both operands.
https://github.com/rchain/rchain/blob/09d9acc916e008dc93a7289202262fd5a4d9d41d/rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala#L491-L496

The same error will be thrown for `if("something" > 0) {}` or `if([] > []) {}` expressions. Because of the bug reported in this issue expressions like `Nil > 0` will not get to this point of operand checking.

I extended my previous comment with the reference to handled operands for both comparison and boolean operators.

**Comment by tgrospic** [comment]
> > Calling x.get on a process x which doesn't support it will throw exception. So the problem is again if x is Nil which trigger this bug before it evaluates get method.
> 
> Yep, that is the reason I assumed.
> `and` branches should be evaluated from left to right. If the left branch returns false, the right branch do not need be evaluated.
> `or` branches are similar, if left one returns true, the right branch do not need be evaluated.
> Agree?

@wangjia184 I agree. Boolean operators should be lazy in evaluation of operands. I created a new issue for that bug #3362. I've also renamed title of this issue to be more specific to the bug expressed here.
